File: advance/src/eu/advance/logistics/flow/engine/api/ds/AdvanceSOAPEndpoint.java
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */

package eu.advance.logistics.flow.engine.api.ds;

import hu.akarnokd.reactive4java.base.Func0;
import hu.akarnokd.utils.xml.XNElement;
import hu.akarnokd.utils.xml.XNSerializable;

import java.net.MalformedURLException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;

import org.slf4j.LoggerFactory;

import eu.advance.logistics.flow.engine.api.core.Copyable;
import eu.advance.logistics.flow.engine.api.core.HasPassword;
import eu.advance.logistics.flow.engine.api.core.Identifiable;

/**
 * Describes a SOAP communication channel.
 * @author akarnokd, 2011.09.20.
 */
public class AdvanceSOAPEndpoint extends AdvanceCreateModifyInfo 
implements XNSerializable, HasPassword, Copyable<AdvanceSOAPEndpoint>, Identifiable<String> {
	/** The name used to reference this channel from blocks. */
	public String name;
	/** The endpoint URL. */
	public URL endpoint;
	/** The target object URI. */
	public URI targetObject;
	/** The target namespace. */
	public URI targetNamespace;
	/** The remote method. */
	public String method;
	/** The communication should be encrypted. */
	public boolean encrypted;
	/** The keystore for the encryption key. */
	public String keyStore;
	/** The key alias for the encryption. */
	public String keyAlias;
	/**
	 * The password.
	 * <p>Note that passwords are never returned from the 
	 * control API calls and are always {@code null}.</p> 
	 * <p>An empty password should be an empty {@code char} array. To keep
	 * the current password, use {@code null}.</p>
	 * @return  
	 */
	private char[] password;
	/**
	 * Create a new instance of the class.
	 */
	public static final Func0<AdvanceSOAPEndpoint> CREATOR = new Func0<AdvanceSOAPEndpoint>() {
		@Override
		public AdvanceSOAPEndpoint invoke() {
			return new AdvanceSOAPEndpoint();
		}
	};
	@Override
	public void load(XNElement source) {
		name = source.get("name");
		try {
			endpoint = new URL(source.get("endpoint"));
		} catch (MalformedURLException ex) {
			LoggerFactory.getLogger(AdvanceSOAPEndpoint.class).error(ex.toString(), ex);
		}
		try {
			targetObject = new URI(source.get("target-object"));
		} catch (URISyntaxException ex) {
			LoggerFactory.getLogger(AdvanceSOAPEndpoint.class).error(ex.toString(), ex);
		}
		try {
			targetNamespace = new URI(source.get("target-namespace"));
		} catch (URISyntaxException ex) {
			LoggerFactory.getLogger(AdvanceSOAPEndpoint.class).error(ex.toString(), ex);
		}
		method = source.get("method");
		encrypted = source.getBoolean("encrypted");
		keyStore = source.get("keystore");
		keyAlias = source.get("keyalias");
		password = getPassword(source, "password");
		super.load(source);
	}
	@Override
	public void save(XNElement destination) {
		destination.set("name", name);
		destination.set("endpoint", endpoint);
		destination.set("target-object", targetObject);
		destination.set("target-namespace", targetNamespace);
		destination.set("method", method);
		destination.set("encrypted", encrypted);
		destination.set("keystore", keyStore);
		destination.set("keyalias", keyAlias);
		setPassword(destination, "password", password);
		super.save(destination);
	}
	@Override
	public AdvanceSOAPEndpoint copy() {
		AdvanceSOAPEndpoint result = new AdvanceSOAPEndpoint();
		result.name = name;
		result.endpoint = endpoint;
		result.targetObject = targetObject;
		result.method = method;
		result.targetNamespace = targetNamespace;
		result.encrypted = encrypted;
		result.keyStore = keyStore;
		result.keyAlias = keyAlias;
		result.password = password != null ? password.clone() : null;
		assignTo(result);
		return result;
	}
	@Override
	public char[] password() {
		return password != null ? password.clone() : null;
	}
	@Override
	public void password(char[] newPassword) {
		password = newPassword != null ? newPassword.clone() : null;
	}
	@Override
	public String id() {
		return name;
	}
}

File: advance-api-cs/AdvanceAPIClient/Classes/AdvanceSOAPEndpoint.cs
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Globalization;
using System.Xml;

using AdvanceAPIClient.Core;
using AdvanceAPIClient.Interfaces;

namespace AdvanceAPIClient.Classes
{
    /// <summary>
    /// Represents the connection information to a POP3(s)/IMAP(s) based email-box to 
    /// </summary>
    public class AdvanceSOAPEndpoint : AdvanceCreateModifyInfo, IPassword, ICopiable<AdvanceSOAPEndpoint>, IIdentifiable<string>
    {
        /// <summary>
        /// Identifier for referencing this box
        /// </summary>
        public string Id { get { return this.Name; } }
        public string Name;
        /// <summary>
        /// user password or key password for authentication
        /// </summary>
        public char[] Password
        {
            get { return (this.password == null) ? null : (char[])this.password.Clone(); }
            set { this.password = (value == null) ? null : (char[])value.Clone(); }
        }
        private char[] password;

        /// <summary>
        /// endpoint URL
        /// </summary>
        public Uri Endpoint;
        /// <summary>
        /// target object URI
        /// </summary>
        public Uri TargetObject;
        /// <summary>
        /// target namespace
        /// </summary>
        public Uri TargetNamespace;
        /// <summary>
        /// remote method
        /// </summary>
        public string Method;
        /// <summary>
        /// If true, communication should be encrypted
        /// </summary>
        public bool IsEncrypted;
        /// <summary>
        /// Keystore for the encryption key
        /// </summary>
        public string KeyStore;
        /// <summary>
        /// Key alias for the encryption
        /// </summary>
        public string KeyAlias;

        public AdvanceSOAPEndpoint() : base() { }

        /// <summary>
        /// Fill object data from xml node
        /// </summary>
        /// <param name="source">Source Xml</param>
        protected override void LoadFromXmlNode(XmlNode source)
        {
            base.LoadFromXmlNode(source);
            base.LoadFromXmlNode(source);
            this.Name = GetAttribute(source, "name");
            this.password = GetPassword(source, "password");
            this.Endpoint = GetUriAttribute(source, "endpoint");
            this.TargetObject = GetUriAttribute(source, "target-object");
            this.TargetNamespace = GetUriAttribute(source, "target-namespace");
            this.Method = GetAttribute(source, "method", null);
            this.IsEncrypted = GetBoolAttribute(source, "encrypted", false);
            this.KeyStore = GetAttribute(source, "keystore");
            this.KeyAlias = GetAttribute(source, "keyalias", null);
        }

        /// <summary>
        /// Fills  object's xml
        /// </summary>
        /// <param name="node">Empty node for the object with name "tagname"</param>
        protected override void FillXmlElement(XmlElement node)
        {
            base.FillXmlElement(node);
            AddAttribute(node, "name", this.Name);
            SetPassword(node, "password", this.password);
            AddAttribute(node, "endpoint", this.Endpoint);
            AddAttribute(node, "target-object", this.TargetObject);
            AddAttribute(node, "target-namespace", this.TargetNamespace);
            AddAttribute(node, "method", this.Method);
            AddBoolAttribute(node, "encrypted", this.IsEncrypted);
            AddAttribute(node, "keystore", this.KeyStore);
            AddAttribute(node, "keyalias", this.KeyAlias);
        }

        public AdvanceSOAPEndpoint Copy()
        {
            AdvanceSOAPEndpoint result = new AdvanceSOAPEndpoint();
            this.AssignTo(result);
            result.Name = this.Name;
            result.password = this.Password;
            result.Endpoint = this.Endpoint;
            result.TargetObject = this.TargetObject;
            result.TargetNamespace = this.TargetNamespace;
            result.Method = this.Method;
            result.IsEncrypted = this.IsEncrypted;
            result.KeyStore = this.KeyStore;
            result.KeyAlias = this.KeyAlias;
            return result;

        }
    }
}
